[Wamika Malik] iP

wamikamalik opened the pull request.

wamikamalik committed:
Add level 0

Greets the user and exits.

wamikamalik committed:
Add Level-1

Echo and exit

wamikamalik committed:
Add tasks handler

Takes in tasks and lists them.

wamikamalik committed:
Add mark as done

Marks the task with a tick marker if done.

wamikamalik committed:
Update Coding Standard

Updated the code to adhere to the coding standard.

wamikamalik committed:
Refactor code

made the code a bit more readable.

wamikamalik committed:
Improve code readability.

refactored the code to make it easier to read.

wamikamalik committed:
Add task types

Updated the code to take in tasks as specific types like event, todo, deadline.

wamikamalik committed:
Add automated text UI testing

wamikamalik committed:
Improve code quality

Lee-Juntong commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,36 @@
+        this.message = message;
+    }
+
+    public CommandType extractType() {
+        String command = message.trim().split(" ")[0];
+        if(command.equalsIgnoreCase("bye")) {
+            return CommandType.EXIT;
+        } else if(command.equalsIgnoreCase("list")) {
+            return CommandType.LIST;
+        } else if(command.equalsIgnoreCase("done")) {
+            return CommandType.MARK_DONE;
+        } else if(message.length()<5) {
+            return CommandType.NONE;
+        } else {
+            if(command.equalsIgnoreCase("todo")){
+                return CommandType.TODO;
+            } else if(command.equalsIgnoreCase("deadline")){
+                return CommandType.DEADLINE;
+            } else if(command.equalsIgnoreCase("event")) {
+                return CommandType.EVENT;
+            } else {
+                return CommandType.NONE;
+            }
+        }
+    }
```
I like this part as this function compiles all possible command the user can input, and then give these tasks to the respective functions.

Lee-Juntong commented on src/main/java/CommandType.java:
```diff
@@ -0,0 +1,9 @@
+public enum CommandType {
+    EXIT,
+    LIST,
+    MARK_DONE,
+    TODO,
+    EVENT,
+    DEADLINE,
+    NONE
+}
```
This is a good practice of enum!

Lee-Juntong commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,206 @@
+     * prints a line of 100 character length.
+     *
+     */
+    public static void printLine() {
+
+        for(int i = 0; i<100; i++) {
+            System.out.print("-");
+        }
+        System.out.println();
+
+    }
```
I like this function as we are using the line everywhere of the iP, extract this as a function is a wise choice. However, maybe 100 '-' is too long?

wamikamalik replied:
Got it! I will try making the line shorter.

Lee-Juntong commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,60 @@
+    /**total number of tasks in the list*/
+    private static int noOfTasks = 0;
+    /**number of incomplete tasks*/
+    public static int noOfIncompleteTasks = 0;
```
Perhaps NoOfIncompleteTasks would be a better name?

wamikamalik replied:
Thank you for the suggestion! Should this variable name be in camel case?

Lee-Juntong reviewed (commented):
Overall, there is almost no coding standard violation, so I believe this could be good code.

Johnson-Yee commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,206 @@
+        while(true) {
+            //the following code interprets the command entered by the user and takes appropriate actions.
+            CommandType type = command.extractType();
+            switch (type) {
+                case EXIT:
+                    System.out.println(BYE_MESSAGE);
+                    break;
+                case LIST:
+                    listTasks(tasks);
+                    break;
+                case MARK_DONE:
+                    int itemNo = extractItemNo(command.getMessage());
+                    executeDone(tasks, itemNo);
+                    break;
+                case TODO:
+                    executeAdd(command.getMessage(), tasks, CommandType.TODO);
+                    break;
+                case DEADLINE:
+                    executeAdd(command.getMessage(), tasks, CommandType.DEADLINE);
+                    break;
+                case EVENT:
+                    executeAdd(command.getMessage(), tasks, CommandType.EVENT);
+                    break;
+                case NONE:
+                    System.out.println(UNIDENTIFIED_MESSAGE);
+            }
```
Just a small indentation error here. There is no indentation for case clauses. https://se-education.org/guides/conventions/java/basic.html :D

wamikamalik replied:
Got it! I seemed to have missed that out. Thank you for pointing out!

Johnson-Yee commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,206 @@
+     */
+    private static void welcomeMessage() {
+
+        String logo = " __A__   _C__   _E__\n"
+                    + "/  _  \\ /  __| |  __|\n"
+                    + "| |_| | | |    | |__\n"
+                    + "|  _  | | |__  | |__\n"
+                    + "| | | | \\____| |____|\n";
+
+        System.out.print(logo);
+        printLine();
+
+        System.out.println("\tHey there! I'm Ace, your very own personal assistant.");
+        System.out.println("\tHow can I help you today?\n");
+        printLine();
```
Could consider changing the function name to printWelcomeMessage()

wamikamalik replied:
Got it! will do that in the next commit.

Johnson-Yee commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,206 @@
+        if(isCorrect) {
+            System.out.println("\n\tAdded task \"" + tasks[Task.getNoOfTasks() - 1] + "\" to your To-Do's!");
+            System.out.println("\tYou now have " + Task.getNoOfIncompleteTasks() + " incomplete task"
+                    + (Task.getNoOfIncompleteTasks() != 1 ? "s" : "") + ".");
```
Could define the magic number here :D

wamikamalik replied:
Thank you for the suggestion! I will try to define an appropriate constant.

Johnson-Yee commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,206 @@
+            }
+            System.out.println();
+        }
+
```
Could remove some uneccessary blank lines

wamikamalik replied:
Yep! I will work on it.

Johnson-Yee commented on src/main/java/CommandType.java:
```diff
@@ -0,0 +1,9 @@
+public enum CommandType {
```
Accurate use of PascalCase! :)

wamikamalik committed:
Add DukeException

basic error handling for empty task description, missing identifier and unidetified command.

wamikamalik committed:
Add NOT_A_NUMBER exception

wamikamalik committed:
Update error handling for add tasks

wamikamalik committed:
Update output test files

wamikamalik committed:
Merge branch 'branch-Level-5'

wamikamalik committed:
Add package duke

wamikamalik committed:
Add multiple packages for duke

packages added include command, task and exception.

wamikamalik committed:
Merge branch 'branch-A-Packages'

MeLoveCarbs commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,226 @@
+    public static final String DEADLINE_IDENTIFIER = "/by";
+    public static final String EVENT_IDENTIFIER = "/on";
+
+    /**
+     * prints a line of 100 character length.
+     *
+     */
+    public static void printLine() {
```
Maybe you can leave out the comment since it is not necessary, the method name is self explanatory. In general, if the method name is self-explanatory, it is okay not to write documentation comments for those :)

wamikamalik replied:
okay! I will look into it! Thanks!

MeLoveCarbs commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,226 @@
+
+        Scanner in = new Scanner(System.in);
+        Command command = new Command(in.nextLine());
+        Task[] tasks = new Task[100];
```
Maybe you can try looking into ArrayList from Java so that you won't have to declare a fixed size array. What happens if user enter more than 100 task?

wamikamalik replied:
Yep! Will do that in the next push.

MeLoveCarbs commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,226 @@
+        while(true) {
+            //the following code interprets the command entered by the user and takes appropriate actions.
+            try {
+                CommandType type = command.extractType();
```
Nice method you have there! :) I can easily identify what type of command it is from the constant it returns

MeLoveCarbs commented on src/main/java/duke/exception/ExceptionType.java:
```diff
@@ -0,0 +1,19 @@
+package duke.exception;
+
+public enum ExceptionType {
```
Very useful enum class so that you Dont Repeat Youself and avoid string literals :) great job

MeLoveCarbs commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,226 @@
+        case TODO:
+            try {
+                String todoName = command.substring(5).trim();
+                tasks[Task.getNoOfTasks()] = new Todo(todoName);
+            } catch (StringIndexOutOfBoundsException e) {
+                throw new DukeException(ExceptionType.MISSING_DESCRIPTION);
+            }
+            break;
+        case DEADLINE:
+            String[] rawName = command.trim().split(DEADLINE_IDENTIFIER);
+            if (rawName[0].trim().equalsIgnoreCase("deadline")) {
+                throw new DukeException(ExceptionType.MISSING_DESCRIPTION);
+            }
+            if (!command.contains(DEADLINE_IDENTIFIER)) {
+                throw new DukeException(ExceptionType.MISSING_IDENTIFIER);
+            }
+            if(rawName.length<2) {
+                throw new DukeException(ExceptionType.MISSING_ON_BY);
+            }
+            String deadlineName = rawName[0].substring(9).trim();
+            String by = rawName[1].trim();
+            tasks[Task.getNoOfTasks()] = new Deadline(deadlineName, by);
+            break;
+        case EVENT:
+            String[] rawEventName = command.trim().split(EVENT_IDENTIFIER);
+            if (rawEventName[0].trim().equalsIgnoreCase("event")) {
+                throw new DukeException(ExceptionType.MISSING_DESCRIPTION);
+            }
+            if (!command.contains(EVENT_IDENTIFIER)) {
+                throw new DukeException(ExceptionType.MISSING_IDENTIFIER);
+            }
+            if(rawEventName.length<2) {
+                throw new DukeException(ExceptionType.MISSING_ON_BY);
+            }
+            String eventName = rawEventName[0].substring(6).trim();
+            String on = rawEventName[1].trim();
+            tasks[Task.getNoOfTasks()] = new Event(eventName, on);
+            break;
+        }
+    }
```
Maybe you can try extracting out certain lines of code to shorten the method? It seems a bit long

wamikamalik replied:
Thanks for pointing out!

wamikamalik committed:
Implement ArrayList<> to store tasks

wamikamalik committed:
Add delete feature

wamikamalik committed:
Update text UI testing

wamikamalik committed:
Refactor code

wamikamalik committed:
add file manger to save tasks

wamikamalik committed:
Reoragnise packages

wamikamalik committed:
Merge branch 'branch-Level-6'

wamikamalik committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Duke.java

wamikamalik committed:
Update error handling for done/delete

wamikamalik committed:
Add class Ui

wamikamalik committed:
Add class TaskList

wamikamalik committed:
Add class Parser

wamikamalik committed:
Update code using new classes

wamikamalik committed:
Update test files

wamikamalik committed:
Refactor code

wamikamalik committed:
Add toDate method to convert dates from one format to another

wamikamalik committed:
Implement toDate() method for tasks

wamikamalik committed:
Update coding standard

wamikamalik committed:
Add code to print tasks due today

wamikamalik committed:
Add find feature

wamikamalik committed:
Add comments to Duke and Ui

wamikamalik committed:
Add comments to parser, exception and file manager classes

wamikamalik committed:
Add comments to task related classes

wamikamalik committed:
Merge pull request #1 from wamikamalik/branch-Level-8

Branch level 8

wamikamalik committed:
Revert "Branch level 8"

wamikamalik committed:
Merge pull request #2 from wamikamalik/revert-1-branch-Level-8

Revert "Branch level 8"

wamikamalik committed:
refactor code

wamikamalik committed:
Merge branch 'master' into branch-Level-8

wamikamalik committed:
Merge pull request #3 from wamikamalik/branch-Level-8

refactor code

wamikamalik committed:
Revert "Revert "Branch level 8""

This reverts commit 7660b789cfd007038af0d7d4fcc28eaadbbdf9c2.

wamikamalik committed:
Merge https://github.com/wamikamalik/ip

# Conflicts:
#	src/main/java/Duke.java

wamikamalik committed:
Merge branch 'master' into branch-Level-9

# Conflicts:
#	src/main/java/Ui.java
#	src/main/java/duke/command/Parser.java
#	src/main/java/duke/exception/ExceptionType.java
#	src/main/java/duke/task/TaskList.java

wamikamalik committed:
Merge branch 'master' into branch-A-JavaDoc

wamikamalik committed:
Merge pull request #4 from wamikamalik/branch-Level-9

Add find feature

wamikamalik committed:
Merge pull request #5 from wamikamalik/branch-A-JavaDoc

Branch a java doc

wamikamalik committed:
Add help feature

wamikamalik committed:
Add user guide

wamikamalik committed:
Update user guide

wamikamalik committed:
Set theme jekyll-theme-cayman

wamikamalik committed:
Add content table to user guide

wamikamalik committed:
Merge https://github.com/wamikamalik/ip

wamikamalik committed:
Update user guide
